src/sinks/datadog_archives.rs

```diff
@@ -0,0 +1,571 @@
+        key,
+        content_encoding: Compression::gzip_default().content_encoding(),
+        options: aws_s3::S3Options {
+            acl: options.acl,
+            grant_full_control: options.grant_full_control,
+            grant_read: options.grant_read,
+            grant_read_acp: options.grant_read_acp,
+            grant_write_acp: options.grant_write_acp,
+            server_side_encryption: options.server_side_encryption,
+            ssekms_key_id: options.ssekms_key_id,
+            storage_class: options.storage_class,
+            tags: options.tags,
+            content_encoding: None,
+            content_type: None,
+        },
+    }
+}
+
+/// Applies the following transformations to align event's schema with DD:
+/// - `_id` is generated in the sink(format described below);
+/// - `date` is set from the Global Log Schema's `timestamp` mapping, or to the current time if missing;
+/// - `message`,`host` are set from the corresponding Global Log Schema mappings;
+/// - `source`, `service`, `status`, `tags` and other reserved attributes are left as is;
+/// - the rest of the fields is moved to `attributes`.
+fn encode_event(event: Event) -> Option<EncodedEvent<PartitionInnerBuffer<Vec<u8>, Bytes>>> {
+    lazy_static! {
+        static ref KEY_PREFIX: Template =
+            Template::try_from("/dt=%Y%m%d/hour=%H/").expect("invalid object key format");
+    }
+    let key = KEY_PREFIX
+        .render_string(&event)
+        .map_err(|error| {
+            emit!(TemplateRenderingFailed {
+                error,
+                field: Some("key_prefix"),
+                drop_event: true,
+            });
+        })
+        .ok()?;
+
+    let mut log = event.into_log();
+
+    log.insert("_id", generate_log_id());
+
+    let timestamp = log
+        .remove(crate::config::log_schema().timestamp_key())
+        .unwrap_or_else(|| chrono::Utc::now().timestamp_millis().into());
+    log.insert(
+        "date",
+        timestamp
+            .as_timestamp()
+            .cloned()
+            .unwrap_or_else(chrono::Utc::now)
+            .to_rfc3339_opts(SecondsFormat::Millis, true),
+    );
+    if let Some(message) = log.remove(crate::config::log_schema().message_key()) {
+        log.insert("message", message);
+    }
+    if let Some(message) = log.remove(crate::config::log_schema().host_key()) {
+        log.insert("host", message);
+    }
+
+    let mut attributes = BTreeMap::new();
+    let custom_attributes: Vec<String> = log
+        .keys()
+        .filter(|path| !RESERVED_ATTRIBUTES.contains(path.as_str()))
+        .collect();
+    for path in custom_attributes {
+        if let Some(value) = log.remove(&path) {
+            attributes.insert(path, value);
+        }
+    }
+    log.insert("attributes", attributes);
+
+    let bytes = serde_json::to_vec(&log)
+        .map(|mut b| {
+            b.push(b'\n');
+            b
+        })
+        .expect("Failed to encode event as json, this is a bug!");
```

Comment:
I think it would be simpler to bump up the expect:
```suggestion
    let mut bytes = serde_json::to_vec(&log)
        .expect("Failed to encode event as json, this is a bug!");
    bytes.push('\n');
```

Reply:
Agree.